### ISSUE TITLE
Remove mock cruft from syntax

### ISSUE THREAD

**Issue by dblock** [dblock/ocmock-expecta#2]
I want this:

``` objc
id mock = [OCMockObject partialMockForObject:ARSwitchBoard.sharedInstance];
[[mock expect] loadPath:@"/fair-profile-id/browse/booths"];
[_fairGuideVC loadAllExhibitors];
[mock verify];
[mock stopMocking];
```

to look like this:

``` objc
expect(ARSwitchBoard.sharedInstance).to.receive(@loadPath).with(@"/fair-profile-id/browse/booths");
[_fairGuideVC loadAllExhibitors];
```

**Comment by dblock** [comment]
cc: @robb, you wanted an example

**Comment by robb** [comment]
So this would work

``` objc
expect(ARSwitchBoard.sharedInstance).to.receive(loadPath:).with(@"/fair-profile-id/browse/booths");

[_fairGuideVC loadAllExhibitors];
```

but this wouldn't

``` objc
[_fairGuideVC loadAllExhibitors];

expect(ARSwitchBoard.sharedInstance).to.receive(loadPath:).with(@"/fair-profile-id/browse/booths");
```

because matchers are evaluated once they leave scope?

**Comment by dblock** [comment]
That would be totally fine. I think expectations should come first.

**Comment by orta** [comment]
OK, so the [dev branch](https://github.com/dblock/ocmock-expecta/tree/dev) has got this:

```
  it(@"checks for a method not being called", ^{
      expect(sut).to.receive(@selector(method));
      [sut method2];
  });

  it(@"checks for a return value", ^{
      expect(sut).receive(@selector(method2)).returning(@2);
      [sut method2];
  });

  it(@"checks for an argument to the method", ^{
      expect(sut).receive(@selector(method3:)).with(@[@"thing"]);
      [sut method3:@"thing"];
  });

```

But it doesn't work, this is because the mock objects are not replacing the originals, so `sut` in this context is the original object and not theOCMock proxy.

I'm a bit stuck here, @segiddins recommended setting a shadow variable. I've had a think about that but I'm not sure if I can have a macro that supports multiple variable tokens
like this:

```
#define expect(...).to.receive(...); 
```

which could grab the whole line and then add a shadow var based on the first one. Any ideas?

**Comment by robb** [comment]
I don't think there is a way to hoist something inside the `expect()` up to its outer scope but this would maybe be a compromise:

``` objc
id sut = foo;

it(@"checks for an argument to the method", ^{
    @mockify(sut) // shadows the existing sut

    expect(sut).receive(@selector(method3:)).with(@[@"thing"]);
    [sut method3:@"thing"];
});
```

**Comment by robb** [comment]
Alternatively, you could use stubs that maintain the identity of `sut`, but I don't know if OCMock supports that.

**Comment by orta** [comment]
I think `@mockify` option is better to me, I can check inside the `expect` that it is a mock and fail on the validation too making it fail sooner, quicker. Also makes having multiple `receives` checks much much easier.

Thanks @robb - I owe you a beer. :beer:

**Comment by orta** [comment]
Done!

**Comment by dblock** [comment]
Maybe you can work @mockify into the expect macro?

On Wednesday, February 4, 2015, Orta notifications@github.com wrote:

> Closed #2 https://github.com/dblock/ocmock-expecta/issues/2.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/dblock/ocmock-expecta/issues/2#event-230416803.

## 

dB. | Moscow - Geneva - Seattle - New York
code.dblock.org - @dblockdotorg http://twitter.com/#!/dblockdotorg -
artsy.net - github/dblock https://github.com/dblock

**Comment by orta** [comment]
Afraid not, that was my problem yesterday. I wouldn't be comfortable changing all `expects` to change your context into a partialMock version, nor is there a way to find out if it's part of a combination with `receive`.
